### ISSUE TITLE
What's the point?!

### ISSUE THREAD

**Issue by jactain** [amplitudesxd/IPHasher#8]
I recently contributed to this project out of boredom and it was good reading how others managed to get around the problem and find a solution that performs well enough. But I really don't understand what's the point of bruteforcing hashed ipv4 addresses instead of pregenerating every possibility (ip;hash) and use that in a key value database so it would be almost instantaneous and no need to wait for bruteforcing.

**Comment by amplitudesxd** [comment]
This was mainly a way to just see the performance of different languages when put up to this task, and as a proof of concept to show how hashing IP's as a form of security, while better than nothing, is still insecure.

While a K/V store would indeed perform better, it requires ~100GB of storage to store each IPv4 address and the corresponding SHA hash. You'd either need 100GB of storage with a fast enough disk, or RAM.

Additionally, locally on my PC, bruteforcing at 65M IPs/sec would perform far better than my drive at 300 MB/s. (dn)

**Comment by Erouax** [comment]
🤣🤣🤣🤣🤣

**Comment by jactain** [comment]
dropbox here i come
![image](https://github.com/amplitudesxd/IPHasher/assets/142063706/4add11f7-d682-4542-ae91-658c99220669)

**Comment by invakid404** [comment]
@jactain On the topic of pregenerating every possibility -- I made that a thing and it is indeed faster than bruteforcing, even on the ancient Toshiba HDD I tested it on. The implementation is in the [db](https://github.com/amplitudesxd/IPHasher/tree/main/db) folder. The pregeneration/indexing step takes a while (took over 24 hours on my end), but the querying is instant afterwards. I could share the database I ended up with, e.g. by making a torrent, but I don't see anyone being interested in downloading 300 gigabytes worth of IPv4 hashes :D I could probably shrink the database if I stored the hashes and IPs as raw bytes instead, but I already spent enough time on this :D

The bottom line is that precomputing and indexing the entire search space is indeed a viable solution for this problem in particular (more viable than @amplitudesxd thought :D), but as mentioned, the goal of this project/repo was to see how much performance could be squeezed out of programming languages when tackling a problem like this and just exploring the various optimizations and approaches.
